src/transport/manager/mod.rs

```diff
@@ -430,157 +449,50 @@ impl TransportManager {
-            peers.insert(
-                peer,
-                PeerContext {
-                    state,
-                    secondary_connection,
-                    addresses,
-                },
-            );
-
-            return Ok(());
-        }
+        let context = peers.entry(peer).or_insert_with(|| PeerContext::default());
 
-        let mut records: HashMap<_, _> = addresses
-            .addresses(limit)
-            .into_iter()
-            .map(|address| (address.clone(), AddressRecord::new(&peer, address, 0, None)))
-            .collect();
+        // Check if dialing is possible before allocating addresses.
+        match context.state.can_dial() {
+            StateDialResult::AlreadyConnected => return Err(Error::AlreadyConnected),
+            StateDialResult::DialingInProgress => return Ok(()),
+            StateDialResult::Ok => {}
+        };
 
-        if records.is_empty() {
+        // The addresses are sorted by score and contain the remote peer ID.
+        // We double checked above that the remote peer is not the local peer.
+        let dial_addresses = context.addresses.addresses(limit);
+        if dial_addresses.is_empty() {
             return Err(Error::NoAddressAvailable(peer));
         }
-
-        let locked_addresses = self.listen_addresses.read();
-        for record in records.values() {
-            if locked_addresses.contains(record.as_ref()) {
-                tracing::warn!(
-                    target: LOG_TARGET,
-                    ?peer,
-                    ?record,
-                    "tried to dial self",
-                );
-
-                debug_assert!(false);
-                return Err(Error::TriedToDialSelf);
-            }
-        }
-        drop(locked_addresses);
-
-        // set connection id for the address record and put peer into `Opening` state
-        let connection_id =
-            ConnectionId::from(self.next_connection_id.fetch_add(1usize, Ordering::Relaxed));
+        let connection_id = self.next_connection_id();
 
         tracing::debug!(
             target: LOG_TARGET,
             ?connection_id,
-            addresses = ?records,
+            addresses = ?dial_addresses,
             "dial remote peer",
         );
 
-        let mut transports = HashSet::new();
-        #[cfg(feature = "websocket")]
-        let mut websocket = Vec::new();
-        #[cfg(feature = "quic")]
-        let mut quic = Vec::new();
-        let mut tcp = Vec::new();
+        let transports = Self::open_addresses(&dial_addresses);
 
-        for (address, record) in &mut records {
-            record.set_connection_id(connection_id);
+        // Dialing addresses will succeed because the `context.state.can_dial()` returned `Ok`.
+        let result = context.state.dial_addresses(
+            connection_id,
+            dial_addresses.iter().cloned().collect(),
+            transports.keys().cloned().collect(),
+        );
+        assert_eq!(result, StateDialResult::Ok);
```

Comment:
Looks dangerous...

Reply:
Yep, have removed the assert-eq, thanks 🙏